enhancements/ingress/haproxy-thread-tuning.md

```diff
@@ -0,0 +1,190 @@
+
+In OpenShift 3.x, users could configure routers to increase the number of
+threads that HAProxy spawns, allowing a single router to handle more
+connections without additional health check overhead. This document proposes to
+allow users to employ a similar strategy, and provide a field within the
+IngressController API to specify the number of threads allocated within router
+pods
+
+### Goals
+
+Provide an API for configuring the number of threads handling connections in
+router pods
+
+### Non-Goals
+
+Expose additional performance tuning parameters available within HAProxy
+
+## Proposal
+
+Add the subresource `threading` to the IngressController API. It will currently
+contain one field, `count`.
+
+```go
+type IngressControllerSpec struct {
+	// ...
+	// Existing fields
+	// ...
+
+	// threading defines parameters for configuring threading options within
+	// routers created under this IngressController. See specific threading
+	// fields for their respective definitions and default values.
+	//
+	// +optional
+	Threading IngressControllerThreading `json:"threading,omitempty"`
+}
+
+type IngressControllerThreading struct {
+	// count defines the number of threads created per router pod. Creating
+	// more threads allows each router pod to handle more connections, at the
+	// cost of more system resources used. If this field is empty, the
+	// IngressController will use a default value of 4 threads.
+	//
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Minimum=4
+	// +optional
+	Count int `json:"count,omitempty"`
+}
+```
+
+When unset, the ingress operator will provision routers with 4 threads,
+matching the existing behavior in OpenShift 4.7.
+
+It would be possible to implement this field as a simple integer within the
+`IngressControllerSpec` as `spec.threadCount` or something similar, however
+there are other tunable threading performance options that HAProxy exposes, and
+it is possible that the ingress controller will be updated to support those
+fields at a later date. If any of those fields are added, it would be
+preferable to include those fields within a single subresource. In order to
+make that potential upgrade less painful, the `spec.threading` subresource will
+be added with only one field, `spec.threading.count`.
+
+### User Stories
+
+#### Story 1
+
+> As a cluster administrator, I want to increase the amount of incoming
+> connections my cluster can handle without configuring IngressController
+> sharding
+
+The administrator can patch their existing ingress controller to increase the
+number of router threads:
+
+```sh
+$ oc patch ingresscontroller/<controller-name> --type=merge -p '{"spec": {"threading": {"count": <new-thread-count>} } }'
+```
+
+New router pods will be rolled out with the updated thread count.
+
+### Risks and Mitigations
+
+#### Increased Resource Usage
+
+It's possible that an administrator could configure routers to have enough
```

Comment:
nit: wording; "to over specify the number of threads"? "to overcommit on the the number of threads"?